scripts/genAltSchedules.ts

```diff
@@ -0,0 +1,266 @@
+ * It then parses it and writes the result to the file ../output/alternates.json
+ *      (It does NOT write to ../output/alternates.js, this must be done manually)
+ * 
+ * Warnings are thrown to terminal via console logs because the school's
+ * iCalendar sucks, a lot. Warnings are automatically fixed and nothing to worry about.
+ * Errors are important and have to be fixed manually in the output json.
+ * 
+ * Most of the code was written by Sean Yen and Yu-Ting Chang (I think)
+ */
+
+/**
+ * Things to note:
+ * On certain alternate schedules, lunch collides with 7th period (the iCalendar is weird): 
+ *      In this case the code automagically fixes it
+ *      In particular, we assume lunch always ends 10 minutes before the next period
+ * PSAT testing day is a strange anomaly
+ */
+
+import fetch from 'node-fetch';
+import {writeFileSync} from 'fs';
+import ical from 'ical';
+import chalk from 'chalk';
+
+
+// Types modified from `../client/src/components/schedule/Periods.tsx`;
+// Go there for client side documentation on the parsed values.
+type PeriodObj = {s: number, e: number};
+type UnparsedPeriodObj = PeriodObj & {n: string}; // Unparsed period objects contain a name field
+type DayObj = {
+    0?: PeriodObj, 1?: PeriodObj, 2?: PeriodObj, 3?: PeriodObj, 4?: PeriodObj, 5?: PeriodObj, 6?: PeriodObj, 7?: PeriodObj, 8?: PeriodObj,
+    B?: PeriodObj, L?: PeriodObj, S?: PeriodObj, P?:PeriodObj
+    // G?: PeriodObj, O?: PeriodObj
+}
+
+// Constants
+const EARLIEST_AM_HOUR = 6
+const HTMLnewlineRegex = /<\/?(p|div|br).*?>|\),? *(?=[A-Z0-9])/g
+const noHTMLRegex = /<.*?>/g
+const noNbspRegex = /&nbsp;/g
+const timeGetterRegex = /\(?(1?[0-9])(?::([0-9]{2}))? *(?:am)? *(?:-|–) *(1?[0-9])(?::([0-9]{2}))? *(noon|pm)?\)?/
+const newLineRegex = /\r?\n/g
+const noNewLineBeforeTimeRegex = /\n\(/g // hack for 2019-09-06 schedule
+const schoolYearStart = new Date(2021, 7, 10)
+const schoolYearEnd = new Date(2022, 5, 3)
+
+const altScheduleRegex = /schedule|extended|lunch/i
+const noSchoolRegex = /holiday|no\s(students|school)|break|development/i
+
+const ErrorConsole = (date: string) => {
+    return `${chalk.bgRedBright.black(`Error`)} on ${chalk.red.underline(date)}`;
+}
+const WarningConsole = (date: string) => {
+    return `${chalk.yellow(`Warning`)} on ${chalk.underline(date)}`;
+}
+
+// Parse an iCal summary and description into an array of `UnparsedPeriodObj`s
+function parseAlternate(summary: string | undefined, description: string | undefined, date: string) { // extra date parameter is only for warnings
+    if (!summary) return;
+
+    // If it's a no school day, return an empty array
+    if (noSchoolRegex.test(summary)) {
+        if (description) return;
+        return [];
+    }
+
+    if (!altScheduleRegex.test(summary)) return;
+    if (!description) return;
+
+    description =
+        '\n' +
+        description
+            .replace(noNewLineBeforeTimeRegex, '(')
```

Comment:
Interesting. Also, is the `\n` prepending necessary? Doesn't that just create an empty string first index after description is split by newlines which is then ignored in the iteration?

Reply:
uhh... I don't remember. I think it's just because some other code stupidity relied on having a newline before every line for consistency or smth, idk